launch/launch/actions/execute_process.py

```diff
@@ -686,6 +702,11 @@ async def __execute_process(self, context: LaunchContext) -> None:
                 ),
             )
 
+        if self.__shutdown_received:
+            # Shutdown received, do not run process
+            return
+        self.__running = True
```

Comment:
@Blast545 why not moving this past the try-block, where it's actually true?

Reply:
It wouldn't be of help in that case ...
It would be true when the condition
```
self.process_details is None or self._subprocess_transport is None
```
is false.

The idea is to check if you are already awaiting `async_execute_process` (or about to await it).

Reply:
If you're past the try-block, you're done awaiting for `async_execute_process` and transport is already not `None`. 

But it's true that there's no guarantee as of whether the protocol has established the connection or not. Which makes wonder. Why does the protocol inject the transport into an action "private" attribute? `ExecuteProcess.__execute_process()` could equally do `self._subprocess_transport = transport` and fill in the PID in process details once it's past the `await`. And set `self._running` accordingly.

Reply:
mmm ... I don't think that works.
Imaging the following scenario:
- __execute_process is called, and it's awaiting for async_execute_process to finish.
- The shutdown event is triggered, as `self.__running` is false it won't do nothing.
- At some point `async_execute_process` will finish, and we're not killing the launched process.

Writing that gave me an idea ... `__execute_process` should check `self.__shutdown` after `async_execute_process` returns!
That solves the problem and is cleaner than the running flag!

Reply:
> The shutdown event is triggered, as self.__running is false it won't do nothing.

If self.__running is false and the completion future is not None, that means we're half-way through an async exec and thus the shutdown has to be deferred. It's not a no-op.

> `__execute_process` should check self.__shutdown after async_execute_process returns!

We could do that, yeah, but we still need to check in `_shutdown_process` whether we can rely on `__execute_process` to shutdown or not. And `__execute_process` would have to call `_shutdown_process()` or a subset of it to get the same behavior (timed signalling and stuff).

Reply:
>  completion future is not None, that means we're half-way through an async exec and thus the shutdown has to be deferred

What completion future is indicating that?

> We could do that, yeah, but we still need to check in _shutdown_process whether we can rely on __execute_process to shutdown or not. And __execute_process would have to call _shutdown_process() or a subset of it to get the same behavior (timed signalling and stuff).

Yes

Reply:
> What completion future is indicating that?

[This one](https://github.com/ros2/launch/pull/427/files#diff-61622adb9c2241797a64d96e6065c414R798). Of course, that's up to where you put early returns, if any.

Personally, I'd rather not plague the implementation with early returns and instead have an early return on `execute()`, a flag to tell the process has started execution (can be completion future existence) and a flag to tell if it's actually running (can be transport availability). On shutdown, if the process didn't start you just don't, if the process is on its way you defer, and if the process is running you kill it. There's little gain in catching it mid-flight after `execute()` completed but `async_execute_process()` hasn't been submitted yet.